src/FESpaces.jl

```diff
@@ -41,6 +41,14 @@ function Base.zero(f::DistributedFESpace)
   FEFunction(f,free_values,isconsistent)
 end
 
+function FESpaces.gather_free_values!(free_vals,f::DistributedFESpace,cell_vals)
```

Comment:
Ok. I think I undertand why these twho functions are needed. `cell_vals`, i.e., the input to `gather_free_values!` is the resulting of calling `(a::DistributedCellDof)(f) = evaluate(a,f)`, is that correct? In such a case we can after gathering, call the `FEFunction` constructors.

Reply:
is that correct?

Reply:
Yes! It is used in the context of applying my transfer operators: 

```
function mul!(y::PVector,op::TransferOperator,x::PVector)
    uh = FEFunction(op.from,x,op.dir_vals_from)
    cell_dof  = get_fe_dof_basis(op.to) 
    cell_vals = cell_dof(uh)
    aux = zero_free_values(op.to)
    gather_free_values!(aux,op.to,cell_vals)
    map_parts(y.owned_values,aux.owned_values) do y, aux
        y .= aux
    end
    return y
end
```

Also, this is the non-elegant solution I told you about. I think I can now revisit it using glues or something more elegant. What do you think?

Reply:
I guess that the main concerns about the current version of this function are two:
* It needs temporary dynamic memory allocation, i.e., `aux` (`PVector`).
* It performs a copy from `aux` to `y`. 
And you would like to come up with an optimized solution that solves these two concerns, am I right? In other words, you would like to modify `y` directly without having to rely on the temporary `aux` vector. 

On the other hand, there is something I quite dont understand here. Which is the distributed data layout  (i..e, owner DoFs + ghost DoFs) corresponding to `x`? The one of the columns of the coefficient matrix assembled in `op.from` or the one of `op.from` itself?

Reply:
`x` and `y` correspond to vectors used in the solver, therefore the distributed data layout corresponds to the assembled matrix. I think this is the same problem we had with the sumfac routines, so it can probably be solved in the same way: Basically the matrix can be assembled with both existing strategies, whereas the `FESpaces` always have ghosts. I'll take a look at it at some point, but in any case it does not affect the methods implemented here.

Reply:
Also, the dynamic mem allocation was just a temporary fix while I found the best way to tackle this. if I had found nothing better i would have made `aux` a cache.

Reply:
> x and y correspond to vectors used in the solver, therefore the distributed data layout corresponds to the assembled matrix.

Ok. If this is the case, how can it be that `x` is used as the vector of free DoF values of `op.from`? Shouldn't this be problematic? See, e.g., how I come around this here: https://github.com/gridap/GridapP4est.jl/blob/generate_mesh_hierarchy/src/GMG/InterGridTransferOperators.jl#L283

Reply:
> I think this is the same problem we had with the sumfac routines, so it can probably be solved in the same way: Basically the matrix can be assembled with both existing strategies, whereas the FESpaces always have ghosts.

A subtle thing here. I am almost sure (although this should be 100% confirmed using code debugging) that the distributed data layout of the rows and columns of the assembled matrix does NOT depend on the strategy used for distributed assembly (i.e., FAR versus SAR). In other words, when using FAR for assembly, you wont have more ghost DoFs on the `x` vector than with SAR. I am not sure if you are aware of that from your comment. (you might be, but just to 100% double check).

Reply:
> if I had found nothing better i would have made aux a cache.

Ok. This is the way to go. Not actually cache, but as a member variable of the matrix data structure itself. We thus would only have to address issue "It performs a copy from aux to y.".

BTW, it seems (to me) that you are re-writing code from `PartitionedArrays.jl` in the current version of your function. Look, e.g., https://github.com/fverdugo/PartitionedArrays.jl/blob/master/src/Interfaces.jl#L1682

Note that the copy! function checks for the set of owner identifiers to be consistent among the two vectors. Otherwise the copy cannot be done, it would be incorrect.

Reply:
Finally, if you use copy!, I think that you may re-use the same function for serial and distributed computations.

Reply:
Ok thanks! I'll have a look at all of this.